freeipa/src/main/java/com/sequenceiq/freeipa/service/multiaz/MultiAzCalculatorService.java

```diff
@@ -20,24 +32,92 @@
+            for (Map.Entry<String, CloudSubnet> entry : environment.getNetwork().getSubnetMetas().entrySet()) {
+                CloudSubnet value = entry.getValue();
+                if (!Strings.isNullOrEmpty(value.getName())) {
+                    subnetAzPairs.put(value.getName(), value.getAvailabilityZone());
+                }
+                if (!Strings.isNullOrEmpty(value.getId())) {
+                    subnetAzPairs.put(value.getId(), value.getAvailabilityZone());
+                }
+            }
+        }
+        return subnetAzPairs;
+    }
+
+    public void calculateByRoundRobin(Map<String, String> subnetAzPairs, InstanceGroup instanceGroup, String variant) {
+        Map<String, Integer> subnetUsage = new HashMap<>();
+        Set<String> subnetIds = collectSubnetIds(instanceGroup);
+        initializeSubnetUsage(subnetAzPairs, subnetIds, subnetUsage);
+        collectCurrentSubnetUsage(instanceGroup, subnetUsage);
+        if (!subnetIds.isEmpty() && multiAzValidator.supportedForInstanceMetadataGeneration(instanceGroup)) {
             for (InstanceMetaData instanceMetaData : instanceGroup.getAllInstanceMetaData()) {
                 if (Strings.isNullOrEmpty(instanceMetaData.getSubnetId())) {
-                    instanceMetaData.setSubnetId(subnetId);
-                    //instanceMetaData.setAvailabilityZone(subnetAzPairs.get(leastUsedSubnetId));
+                    Integer numberOfInstanceInASubnet = searchTheSmallestInstanceCountForSubnets(subnetUsage);
+                    String leastUsedSubnetId = searchTheSmallestUsedSubnetID(subnetUsage, numberOfInstanceInASubnet);
+
+                    instanceMetaData.setSubnetId(leastUsedSubnetId);
+                    instanceMetaData.setAvailabilityZone(subnetAzPairs.get(leastUsedSubnetId));
+
+                    subnetUsage.put(leastUsedSubnetId, numberOfInstanceInASubnet + 1);
+                }
+            }
+        }
+    }
+
+    private Integer searchTheSmallestInstanceCountForSubnets(Map<String, Integer> subnetUsage) {
+        return Collections.min(subnetUsage.values());
+    }
+
+    private String searchTheSmallestUsedSubnetID(Map<String, Integer> subnetUsage, Integer currentNumber) {
+        return subnetUsage.entrySet()
+                .stream()
+                .filter(e -> e.getValue().equals(currentNumber))
+                .findFirst()
+                .get()
+                .getKey();
+    }
+
+    private void initializeSubnetUsage(Map<String, String> subnetAzPairs, Set<String> subnetIds, Map<String, Integer> subnetUsage) {
+        for (String subnetId : subnetAzPairs.keySet()) {
+            if (subnetIds.contains(subnetId)) {
+                subnetUsage.put(subnetId, 0);
+            }
+        }
+    }
+
+    private void collectCurrentSubnetUsage(InstanceGroup instanceGroup, Map<String, Integer> subnetUsage) {
+
+        for (InstanceMetaData instanceMetaData : instanceGroup.getAllInstanceMetaData()) {
```

Comment:
After a bit of thinking, I am not sure anymore if this will make any difference. In `core` it was essential for sake of upscale & repair to be able to cope with instances that got terminated / deleted previously (either as part of a downscale, or manual terminatation on the provider side). I wonder what the repair story is for FreeIPA.